core/src/main/java/org/apache/iceberg/jdbc/JdbcTableOperations.java

```diff
@@ -0,0 +1,218 @@
+        String oldMetadataLocation = base.metadataFileLocation();
+        // Start atomic update
+        updateTable(newMetadataLocation, oldMetadataLocation);
+      } else {
+        // table not exists create it!
+        createTable(newMetadataLocation);
+      }
+
+    } catch (SQLIntegrityConstraintViolationException e) {
+      throw new AlreadyExistsException(e, "Table already exists! maybe another process created it!");
+    } catch (SQLTimeoutException e) {
+      throw new UncheckedSQLException("Database Connection timeout!", e);
+    } catch (SQLTransientConnectionException | SQLNonTransientConnectionException e) {
+      throw new UncheckedSQLException("Database Connection failed!", e);
+    } catch (DataTruncation e) {
+      throw new UncheckedSQLException("Database data truncation error!", e);
+    } catch (SQLWarning e) {
+      throw new UncheckedSQLException("Database warning!", e);
+    } catch (SQLException e) {
+      throw new UncheckedSQLException("Failed to connect to database!", e);
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted during commit", e);
+    }
+  }
+
+  private void updateTable(String newMetadataLocation, String oldMetadataLocation)
+      throws SQLException, InterruptedException {
+    int updatedRecords = connections.run(conn -> {
+      try (PreparedStatement sql = conn.prepareStatement(JdbcUtil.DO_COMMIT_SQL)) {
+        // UPDATE
+        sql.setString(1, newMetadataLocation);
+        sql.setString(2, oldMetadataLocation);
+        // WHERE
+        sql.setString(3, catalogName);
+        sql.setString(4, JdbcUtil.namespaceToString(tableIdentifier.namespace()));
+        sql.setString(5, tableIdentifier.name());
+        sql.setString(6, oldMetadataLocation);
+        return sql.executeUpdate();
+      }
+    });
+
+    if (updatedRecords == 1) {
+      LOG.debug("Successfully committed to existing table: {}", tableIdentifier);
+    } else {
+      throw new CommitFailedException("Failed to commit the table %s from catalog %s! " +
+          "Maybe another process changed it!", tableIdentifier, catalogName);
+    }
+
+  }
+
+  private void createTable(String newMetadataLocation) throws SQLException, InterruptedException {
+    int insertRecord = connections.run(conn -> {
+      try (PreparedStatement sql = conn.prepareStatement(JdbcUtil.DO_COMMIT_CREATE_SQL)) {
+        sql.setString(1, catalogName);
+        sql.setString(2, JdbcUtil.namespaceToString(tableIdentifier.namespace()));
+        sql.setString(3, tableIdentifier.name());
+        sql.setString(4, newMetadataLocation);
+        return sql.executeUpdate();
+      }
+    });
+
+    if (insertRecord == 1) {
+      LOG.debug("Successfully committed to new table: {}", tableIdentifier);
+    } else {
+      throw new CommitFailedException("Failed to commit the table %s from catalog %s", tableIdentifier, catalogName);
```

Comment:
should indicate this is a create table operation, maybe message can be `fail to create the table ...`. Same for the update table exception above.